### PR TITLE
Windows CI build

### PR DESCRIPTION
**Pull request by TomMelt**
This PR adds windows CI running on latest version of windows with intel compilers version `2023.2`.

For green computing reasons I have only prepared an intel build. My plan is that we have:
- `gcc` on linux
- `clang` on mac; and
-  `intel` on windows

That way we can get nice coverage without hammering compute resources.

### CODE REVIEW

**Review by jatkinson1000 — COMMENTED** [review]
Thanks @TomMelt
Looks sensible to me (as much as any windows code does...)

I think we need some documentation adding about how to run the test suite for Windows users (definitely online and maybe README, though perhaps online only to avoid pollution)

I know nothing about bat scripts, but is there a reason we apply it in a for loop for each example as part of the workflow rather than running all examples from the bat script as we do in `run_integration_tests.sh`?

File: .github/workflows/test_suite_windows.yml [hunk 1]
```diff
@@ -0,0 +1,58 @@
+name: TestSuiteWindows
+
+on:
+  push:
+    branches: [ "main" ]
+  pull_request:
```
Comment (jwallwork23) [inline .github/workflows/test_suite_windows.yml:6]: Please could you add comments consistently as in the other workflows? My recent workflow PR made them all similar in this respect.

File: .github/workflows/test_suite_windows.yml [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+          cmake --install build
+
+      - name: run tests
+        shell: cmd
+        run: |
+          cd src
+          set PATH=C:\hostedtoolcache\windows\Python\3.12.7\x64\Lib\site-packages;%PATH%
+          set PATH=C:\Program Files (x86)\FTorch\bin;%PATH%
+          set PATH=C:\hostedtoolcache\windows\Python\3.12.7\x64\Lib\site-packages\torch\lib;%PATH%
+          for /d %%i in (1_SimpleNet 2_ResNet18 4_MultiIO) do ..\run_integration_tests.bat build\test\examples\%%i
```
Comment (jwallwork23) [inline .github/workflows/test_suite_windows.yml:58]: Same question as @jatkinson1000 - can this loop not go inside the `run_integration_tests.bat` script?

File: run_integration_tests.bat [hunk 3]
```diff
@@ -0,0 +1,4 @@
+pushd %1
+ctest
+if %errorlevel% neq 0 exit /b %errorlevel%
+popd
```
Comment (jwallwork23) [inline run_integration_tests.bat:4]: This file needs some comments at the top to say what it is (i.e., that it's the Windows version of `run_integration_tests.sh`.

**Review by jwallwork23 — CHANGES_REQUESTED** [review]
Thanks for putting in the effort to get this working @TomMelt. Just a few minor comments from me.

File: .github/workflows/test_suite_windows.yml
Comment (jwallwork23) [inline .github/workflows/test_suite_windows.yml:1]: Perhaps rename the other one to `test_suite_ubuntu.yml` for clarity. We could use reusable workflows to make it easier if we ever want to add another test workflow. I can look into this as a follow-up.

Comment (jwallwork23) [reply .github/workflows/test_suite_windows.yml:1]: Opened #201 and self-assigned.

File: .github/workflows/test_suite_windows.yml (hunk 1 above)
Comment (TomMelt) [reply .github/workflows/test_suite_windows.yml]: fixed by [d177685](https://github.com/Cambridge-ICCS/FTorch/pull/168/commits/d177685da3263399d49e3b0deb2c57d4dcf66fde)

File: .github/workflows/test_suite_windows.yml (hunk 2 above)
Comment (TomMelt) [reply .github/workflows/test_suite_windows.yml]: fixed by [a6ec9c2](https://github.com/Cambridge-ICCS/FTorch/pull/168/commits/a6ec9c2ef7d8e880b5d0efabee8737d24ff5f3e1)

**Comment by TomMelt** [comment]
> I know nothing about bat scripts, but is there a reason we apply it in a for loop for each example as part of the workflow rather than running all examples from the bat script as we do in `run_integration_tests.sh`?

fixed by [a6ec9c2](https://github.com/Cambridge-ICCS/FTorch/pull/168/commits/a6ec9c2ef7d8e880b5d0efabee8737d24ff5f3e1)

File: .github/workflows/test_suite_windows.yml [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+# Controls when the workflow will run
+on:
+  # Triggers the workflow on pushes to the "main" branch, i.e., PR merges
+  push:
+    branches: [ "main" ]
+
+  # Triggers the workflow on pushes to open pull requests with code changes
+  pull_request:
+    paths:
+      - '.github/workflows/test_suite.yml'
```
Comment (jwallwork23) [inline .github/workflows/test_suite_windows.yml:13]: This should be specific to the `test_suite_windows.yml` config.

Similar change in `test_suite_ubuntu.yml` if it's renamed as such.

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -102,6 +102,12 @@ install(FILES "${CMAKE_BINARY_DIR}/modules/ftorch_test_utils.mod"
 
 # Build integration tests
 if(CMAKE_BUILD_TESTS)
+
+  if (WIN32)
```
[the remark this answers is not in the document]
Comment (TomMelt) [reply src/CMakeLists.txt:105]: resolved by #203

File: run_integration_tests.bat (hunk 3 above)
Comment (TomMelt) [reply run_integration_tests.bat]: resolved by b38397a

**Review by jwallwork23 — APPROVED** [review]
Looks good!

File: src/test/README.md [hunk 6]
```diff
@@ -24,8 +24,11 @@ which tests passed and which failed for your build. Note that the examples have
...
-```
-This will automatically install any additional dependencies for the examples.
+Alternatively, run the helper script in the root FTorch directory. Depending on
+which operating system you are running, you will need:
+
+- `./run_integration_tests.sh` for unix (mac and linux)
+- `run_integration_tests.bat` for windows
+
+This will automatically install any additional dependencies for the examples and
+run the tests.
```
Comment (jatkinson1000) [inline src/test/README.md:34]: I wonder if we should clarify _Python_ dependencies (don't think there are any others??) and so it is advisable to be in a venv?

**Review by jatkinson1000 — CHANGES_REQUESTED** [review]
Thanks @TomMelt This looks great, I think we just need the online documentation updating to match (in `pages/` - definitely `testing.md`, but worth a quick scan of anywhere else, including the Windows specific docs).

**Comment by TomMelt** [comment]
> Thanks @TomMelt Looks sensible to me (as much as any windows code does...)
> 
> I think we need some documentation adding about how to run the test suite for Windows users (definitely online and maybe README, though perhaps online only to avoid pollution)
> 
> I know nothing about bat scripts, but is there a reason we apply it in a for loop for each example as part of the workflow rather than running all examples from the bat script as we do in `run_integration_tests.sh`?

solved by #203 and https://github.com/Cambridge-ICCS/FTorch/commit/b38397a8f3c7311fc1466f4490dd1d34665e2acb

File: src/test/README.md (hunk 6 above)
Comment (TomMelt) [reply src/test/README.md]: resolved by 4a23616

File: src/CMakeLists.txt [hunk 7]
```diff
@@ -101,6 +101,10 @@ install(FILES "${CMAKE_BINARY_DIR}/modules/ftorch_test_utils.mod"
 
 # Build integration tests
 if(CMAKE_BUILD_TESTS)
+
+  set (Python3_FIND_VIRTUALENV FIRST)
```
Comment (jatkinson1000) [inline src/CMakeLists.txt:105]: This is a CMake 3.15 feature, so perhaps we need to bump the dependency across FTorch to CMake 3.15?

I think this is still sufficiently old not to be an issue for any users.

**Review by jatkinson1000 — CHANGES_REQUESTED** [review]
Thanks @TomMelt 

This is really great, and a significant hurdle - hopefully you learnt something useful, even if it is at the cost of becoming the "Windows guy". ;P 

The blocker for me is not only just to bump the version of CMake required by FTorch to be 3.15 (for use of the Python VIRTUALENV) feature (see specific comment).
Consider also updating in the example CMakeLists for consistency, and in the documentation. From past experience the easiest way to do this is to grep the full source for `3.5`.

Once that is done I am happy to approve.

Other question is whether there is anything you feel is particularly useful or unintuitive to highlight for Windows users within the workflow. Otherwise we can just update the Windows docs piece by piece as people raise things in future.

File: src/CMakeLists.txt (hunk 7 above)
Comment (TomMelt) [reply src/CMakeLists.txt:105]: done in https://github.com/Cambridge-ICCS/FTorch/pull/168/commits/17d8db2589376ee4c0e5d92c9f99c38cba5720ba

**Comment by TomMelt** [comment]
> Thanks @TomMelt
> 
> This is really great, and a significant hurdle - hopefully you learnt something useful, even if it is at the cost of becoming the "Windows guy". ;P
> 
> The blocker for me is not only just to bump the version of CMake required by FTorch to be 3.15 (for use of the Python VIRTUALENV) feature (see specific comment). Consider also updating in the example CMakeLists for consistency, and in the documentation. From past experience the easiest way to do this is to grep the full source for `3.5`.
> 
> Once that is done I am happy to approve.
> 
> Other question is whether there is anything you feel is particularly useful or unintuitive to highlight for Windows users within the workflow. Otherwise we can just update the Windows docs piece by piece as people raise things in future.

Thanks @jatkinson1000 , I have added a [concrete example](https://github.com/Cambridge-ICCS/FTorch/pull/168/commits/e7ebd4d9db1d8505566f574fb4a00840620db9da) to the online docs just in case anyone has difficulty (most likely my future self)

**Comment by TomMelt** [comment]
> Thanks @TomMelt I think README.md and /pages/cmake.md still need the CMake versioin updating from 3.5 to 3.15.

done in https://github.com/Cambridge-ICCS/FTorch/pull/168/commits/df5340a595e9d4dad767aac2631c4d7e9ed339ca

**Review by jatkinson1000 — APPROVED** [review]
I'm happy with this now.

Thanks @TomMelt it has been a mammoth task you've done here!

**Pull request merged**